engine/crates/composition/src/emit_federated_graph.rs

```diff
@@ -244,7 +245,7 @@ fn emit_fields<'a>(
         let mut end_field_id = None;
 
         if let federated::Definition::Object(id) = definition {
-            let object_name = ctx.out.objects[id.0].name;
+            let object_name = ctx.out.through(id).view(|obj| obj.type_definition_id).name;
```

Comment:
Been staring at for 5 minutes or so, but not sure I get this through/view stuff...

Reply:
Hehe that's not a good sign. The `.view()` methods are intended to return a `View`, that is basically a record struct + its id. The `through` methods are just a shortcut (with the `NestedView` helper) to avoid having to write `graph.view(graph.view(interface_id).type_definition_id)`.

And the `.str()` methods are a shortcut to avoid having to write: `let interface_name = &graph[graph.view(graph.view(interface_id).type_definition_id).name]`. Which is a line I wrote in this PR before this abstraction, but that I found disgusting.

Reply:
We could unify `View` and `NestedView` if we made `View` more like the other abstractions we have and added a `graph: &FederatedGraph` field to it.

Reply:
Maybe "follow" is a better name than "through". Basically take an id, dereference the FederatedGraph with it and the next view in the chain is whatever got dereferenced.

Reply:
I think we could maybe decouple the closure part and make it its own method, have something more orthogonal. I'll give that a try and see if it ends up more readable.

Reply:
I rearranged the methods on View & co a bit so instead of `view()`, `through()` and `str()` with different signatures between FederatedGraph and NestedView, we only have `FederatedGraph::at()` and `NestedView::then()`, plus some more `Deref` magic.

The change is all in this commit: https://github.com/grafbase/grafbase/pull/2180/commits/b6f8cbf094a68fa8b78f8110b96676af297636a5